lib/graphql/schema.rb

```diff
@@ -545,6 +562,9 @@ def rebuild_artifacts
         traversal = Traversal.new(self)
         @types = traversal.type_map
         @instrumented_field_map = traversal.instrumented_field_map
+        @type_reference_map = traversal.type_reference_map
+        @unions = traversal.unions
+        @union_memberships = {}
```

Comment:
I think we should try to populate `@union_memberships` _here_ instead of gradually growing it at runtime. 

The reason is because of how it will behave in threaded and forked environments. In a forking server, every fork will end up with its own copy of `@union_memberships` since each fork will be writing to it, it's not the end of the world, but it's not what we want. In threaded environments, `@union_memberships` becomes shared mutable state, since different threads may be adding to it at different times. In CRuby, the GIL keeps us safe, but in JRuby, this can raise an error.

So, do you think it's possible to pre-fill this hash now instead of fill-as-we-go?

Reply:
ah good catch! will fix.